homeassistant/components/imap/coordinator.py

```diff
@@ -110,7 +106,18 @@ def __init__(self, hass: HomeAssistant, imap_client: IMAP4_SSL) -> None:
 
     async def _async_update_data(self) -> int | None:
         """Update the number of unread emails."""
-        return await self._async_fetch_number_of_messages()
+        try:
+            return await self._async_fetch_number_of_messages()
+        except (
+            UpdateFailed,
+            InvalidAuth,
+            InvalidFolder,
+            AioImapException,
+            asyncio.TimeoutError,
+        ) as ex:
+            self.async_set_update_error(ex)
+            await self._cleanup()
+            return None
```

Comment:
We catch the error here so we can do a proper cleanup and return `None`.

Reply:
We should re-raise `InvalidFolder , `AioImapException `as `UpdateFailed` and `InvalidAuth` as `ConfigEntryAuthFailed` here as it will take care of setting the error and marking it unavailable or trigger reauth

Reply:
Thnx, I have made some changes